src/dbnode/storage/shard.go

```diff
@@ -2650,6 +2658,78 @@ func (s *dbShard) Repair(
 	return repairer.Repair(ctx, nsCtx, nsMeta, tr, s)
 }
 
+func (s *dbShard) AggregateTiles(
+	ctx context.Context,
+	reader fs.DataFileSetReader,
+	sourceNsID ident.ID,
+	sourceBlockStart time.Time,
+	sourceShard databaseShard,
+	opts AggregateTilesOptions,
+	wOpts series.WriteOptions,
+) (int64, error) {
+	latestSourceVolume, err := sourceShard.latestVolume(sourceBlockStart)
+	if err != nil {
+		return 0, err
+	}
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   sourceNsID,
+			Shard:       sourceShard.ID(),
+			BlockStart:  sourceBlockStart,
+			VolumeIndex: latestSourceVolume,
+		},
+		FileSetType: persist.FileSetFlushType,
+		//TODO add after streaming supported - OrderByIndex: true
+	}
+	if err := reader.Open(openOpts); err != nil {
+		return 0, err
+	}
+	defer reader.Close()
+
+	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
+	bytesReader := bytes.NewReader(nil)
+	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
+	var lastWriteError error
+	var processedBlockCount int64
+
+	for {
+		id, tags, data, _, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return processedBlockCount, err
+		}
+
+		data.IncRef()
+		bytesReader.Reset(data.Bytes())
+		dataPointIter.Reset(bytesReader, nil)
+
+		for dataPointIter.Next() {
+			dp, unit, annot := dataPointIter.Current()
+			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
```

Comment:
Do we want to downsample points to the new NS resolution here?

Reply:
This was the flow before we had downsampling implemented.